bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/Bookie.java

```diff
@@ -441,77 +335,122 @@ private void checkIfDirsOnSameDiskPartition(List<File> dirs) throws DiskPartitio
-            if (!newEnv) {
-                for (Cookie journalCookie: journalCookies) {
-                    masterCookie.verifyIsSuperSet(journalCookie);
-                }
-            }
 
+            // 3. read the cookie from registration manager. it is the `source-of-truth` of a given bookie.
+            //    if it doesn't exist in registration manager, this bookie is a new bookie, otherwise it is
+            //    an old bookie.
+            final Versioned<Cookie> rmCookie = readAndVerifyCookieFromRegistrationManager(
+                masterCookie,
+                conf,
+                rm
+            );
+            final boolean newEnv = null == rmCookie;
 
-            for (File dir : allLedgerDirs) {
-                checkDirectoryStructure(dir);
-                try {
-                    Cookie c = Cookie.readFromDirectory(dir);
-                    masterCookie.verifyIsSuperSet(c);
-                } catch (FileNotFoundException fnf) {
-                    missedCookieDirs.add(dir);
-                }
+            // 4. check if the cookie appear in all the directories.
+            List<File> missedCookieDirs = new ArrayList<>();
+            List<Cookie> existedCookies = Lists.newArrayList();
+            if (!newEnv) {
+                existedCookies.add(rmCookie.getValue());
             }
 
+            // 4.1 verify the cookies in journal directories
+            Pair<List<File>, List<Cookie>> journalResult =
+                verifyAndGetMissingDirs(masterCookie, conf, journalDirectories);
+            missedCookieDirs.addAll(journalResult.getLeft());
+            existedCookies.addAll(journalResult.getRight());
+            // 4.2. verify the cookies in ledger directories
+            Pair<List<File>, List<Cookie>> ledgerResult =
+                verifyAndGetMissingDirs(masterCookie, conf, allLedgerDirs);
+            missedCookieDirs.addAll(ledgerResult.getLeft());
+            existedCookies.addAll(ledgerResult.getRight());
+
+            // 5. if it is not a new environment, find out the real directories that miss cookie data
             if (!newEnv && missedCookieDirs.size() > 0) {
                 // If we find that any of the dirs in missedCookieDirs, existed
                 // previously, we stop because we could be missing data
                 // Also, if a new ledger dir is being added, we make sure that
                 // that dir is empty. Else, we reject the request
                 Set<String> existingLedgerDirs = Sets.newHashSet();
-                for (Cookie journalCookie : journalCookies) {
-                    Collections.addAll(existingLedgerDirs, journalCookie.getLedgerDirPathsFromCookie());
+                for (Cookie cookie : existedCookies) {
+                    Collections.addAll(existingLedgerDirs, cookie.getLedgerDirPathsFromCookie());
                 }
                 List<File> dirsMissingData = new ArrayList<File>();
                 List<File> nonEmptyDirs = new ArrayList<File>();
```

Comment:
rename to nonEmptyNewDirs

Reply:
I would prefer not changing names for existing names at the same patch. If other people cherry-pick this change, people will know there is no name changing in existing logic.